src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java

```diff
@@ -125,19 +125,22 @@ private void handleRateLimitSuccess(String deploymentId) {
             return;
         }
 
-        ApiKeyData proxyApiKeyData = new ApiKeyData();
-        context.setProxyApiKeyData(proxyApiKeyData);
-        ApiKeyData.initFromContext(proxyApiKeyData, context);
-        proxy.getApiKeyStore().assignApiKey(proxyApiKeyData);
-
-        proxy.getTokenStatsTracker().startSpan(context);
-
-        context.getRequest().body()
-                .onSuccess(body -> proxy.getVertx().executeBlocking(() -> {
-                    handleRequestBody(body);
-                    return null;
-                }))
-                .onFailure(this::handleRequestBodyError);
+        proxy.getVertx().executeBlocking(() -> {
```

Comment:
maybe it's better to rewrite below code to highlight 2 steps explicitly
```proxy.getVertx().executeBlocking(() -> {
                    ApiKeyData proxyApiKeyData = new ApiKeyData();
                    context.setProxyApiKeyData(proxyApiKeyData);
                    ApiKeyData.initFromContext(proxyApiKeyData, context);
                    proxy.getApiKeyStore().assignApiKey(proxyApiKeyData);
                    proxy.getTokenStatsTracker().startSpan(context);
                    return null;
                })
                .onFailure(this::handleError)
                .map(mapper -> context.getRequest().body()
                        .onSuccess(body -> proxy.getVertx().executeBlocking(() -> {
                            handleRequestBody(body);
                            return null;
                        }))
                        .onFailure(this::handleRequestBodyError));
```

Reply:
I might be wrong. But this code leads to the same error might be processed twice:
.onFailure(this::handleError) and onFailure(this::handleRequestBodyError)).
Doesn't it?
I will take some time the code can be improved.

Reply:
map function only called if future succeeded

Reply:
you're right the function map is called when the future is succeeded but the callback onFailure is called when an error happens in the chain.
For example,
`assign api key is failed -> onFailure is called -> map is not called so no new feature is created -> onFailure is called again`

Reply:
fixed